locator/abstract_replication_strategy.cc

```diff
@@ -68,6 +68,12 @@ inet_address_vector_replica_set abstract_replication_strategy::get_natural_endpo
     return res->second;
 }
 
+bool abstract_replication_strategy::in_natural_endpoints(const noncopyable_function<bool(const inet_address_vector_replica_set&)>& func, const token& search_token, const effective_replication_map& erm) const {
```

Comment:
We need to document what this function really does. The name in_natural_endpoints looks as if we can checking if a node is part of the natrual endpoints set. 

How about something like this:

get_natural_endpoints_and_traverse()

or 

get_natural_endpoints_and_run()

which reflects the fact we are getting the natural endpoints first and then run a function to against the natural endpoints.

Reply:
The point here is that we don't get the natural endpoints like the current get_natural_endpoints that returns a copy of the natural endpoints vector, but rather we're passing a const reference to the endpoints to the given function.  So how about `invoke_on_natural_endpoints`?

Reply:
I raise `for_each_natural_endpoints()`.

Reply:
I also initially thought of "for each" but it would make sense only if the function would have been invoked for each endpoint separately.
But we need access to the vector, e.g. to match only the first entry, which is the primary replica.

Reply:
invoke_on_natural_endpoints is better than my version.   

Regarding the return value, does this make more sense?

void invoke_on_natural_endpoints([&invoke_result]  (const inet_address_vector_replica_set&) {
}, ... )

The return state belongs to the lambda we passed to.

Reply:
Sure. That makes sense, and I also contemplated about this myself.
It's a little harder to use this way since we'll need to keep a variable to hold the result, but that's not terrible.
Ideally we could define it as a template that gets the return type from the functor, but then the implementation will need to be in the header file and it's not worth the extra complication as this isn't called on the hot data path.

Reply:
OK. I vote for no template as well ;)

Reply:
BTW, the `bool` return type is related to why I called this function `in_natural_endpoints` since the function passed to it was supposed to search something in the natural endpoints and to return true iff the condition was found.
But I agree there could be other uses for it that may return something else, like one of the endpoints.

Reply:
Maybe the template isn't that bad.  I'll take a stab at it...
The caller's code would be much more elegant if invoke_on_... just returns the value the function param returns.

Reply:
```
'virtual' cannot be specified on member function templates (fix available)clang(virtual_member_function_template)
```
:-(